Delay setting terminal Job conditions until all pods are terminal

mimowo opened the pull request:
#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #123775

#### Special notes for your reviewer:

Experiments to measure the Job termination times in the e2e test: https://github.com/kubernetes/kubernetes/pull/125816

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Delay setting terminal Job conditions until all pods are terminal.

Additionally, the FailureTarget condition is also added to the Job object in the first Job
status update as soon as the failure conditions are met (backoffLimit is exceeded, maxFailedIndexes, 
or activeDeadlineSeconds is exceeded).

Similarly, the SuccessCriteriaMet condition is added in the first update as soon as the expected number
of pod completions is reached.

Also, introduce the following validation rules for Job status when JobManagedBy is enabled:
1. the count of ready pods is less or equal than active
2. when transitioning to terminal phase for Job, the number of terminating pods is 0
3. terminal Job conditions (Failed and Complete) should be preceded by adding the corresponding interim conditions: FailureTarget and SuccessCriteriaMet
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs
```

mimowo commented:
/test pull-kubernetes-node-e2e-containerd
Ticketed already: https://github.com/kubernetes/kubernetes/issues/122132

mimowo reviewed (commented):
/assign @alculquicondor @atiratree 
I think it is ready for early feedback. PTAL

dejanzele commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -2265,6 +2266,12 @@ func TestSyncJobPastDeadline(t *testing.T) {
+					Type:    batch.JobFailureTarget,
+					Status:  v1.ConditionTrue,
+					Reason:  batch.JobReasonDeadlineExceeded,
+					Message: "Job was active longer than specified deadline",
```
We don't have consts for reasons, right?

Might make sense to introduce them as this messages are repeated in a couple of places across Job controller (in a followup PR though)

alculquicondor replied:
I would only add a const if we use it in production code more than once.
If it's just to avoid repeating it in test cases, I'm not too in favor of it.

mimowo replied:
I agree, I don't think the const for message is needed, but even if we conclude it is useful, I wouldn't change the testing approach in this PR.

dejanzele commented on test/e2e/apps/job.go:
```diff
@@ -718,10 +718,13 @@ done`}
 
 		ginkgo.By("Creating a job")
 		job := e2ejob.NewTestJob("notTerminate", "exceed-active-deadline", v1.RestartPolicyNever, parallelism, completions, &activeDeadlineSeconds, backoffLimit)
+		// set low to TerminationGracePeriodSeconds to ensure quick pod termination
```
```suggestion
		// set TerminationGracePeriodSeconds to a low value to ensure quick pod termination
```

mimowo replied:
Actually, it turned out, in my experiments that this does not help . This is only useful to evict quickly a pod that handles SIGTERM and does not make progress. In this case the terminating pods as mostly in Pending phase. When they enter the Running phase they terminate quickly (seconds)

dejanzele commented on test/e2e/apps/job.go:
```diff
@@ -718,10 +718,13 @@ done`}
 		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
 		ginkgo.By("Ensuring job past active deadline")
-		err = waitForJobFailure(ctx, f.ClientSet, f.Namespace.Name, job.Name, time.Duration(activeDeadlineSeconds+15)*time.Second, "DeadlineExceeded")
+		err = waitForJobFailure(ctx, f.ClientSet, f.Namespace.Name, job.Name, time.Duration(activeDeadlineSeconds+20)*time.Second, "DeadlineExceeded")
```
This was increased to prevent flakiness? Should it have a TODO?

mimowo replied:
Right, I run experiments, here and in https://github.com/kubernetes/kubernetes/pull/125816 by looping the test. I noticed that pod termination may take up to 40s in some cases. I bump it tentatively to 60s.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -2162,6 +2162,7 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
 					jobCtx.podsWithDelayedDeletionPerIndex = getPodsWithDelayedDeletionPerIndex(logger, jobCtx)
 				}
 			}
+			jobCtx.terminating = ptr.To(controller.CountTerminatingPods(jobCtx.activePods))
```
No test cases were depending on this field? Can you add some?

mimowo replied:
This code actually was just to ensure that the test execution does not panic when jobCtx.terminating is nil. Now I fixed the setup for the field, and added test cases which rely on the value, depending on the feature gates enablement.

alculquicondor commented on test/integration/job/job_test.go:
Worth adding an integration test when we first have running Pods and we see only the interim condition. Then we finish the Pods and we see the terminal condition.

mimowo replied:
I've added a dedicated integration test `TestDelayTerminalPhaseCondition` with test cases for (`JobPodReplacementPolicy` and `JobManagedBy` disabled, or any of them enabled). There are also test cases for one success and one failure scenario.

The integration tests are relatively  heavy so covering every scenario might be excessive, so to cover every scenario I also extended the unit tests.

For e2e I would propose that we update the helper functions (e2ejob.WaitForJobFailed and e2ejob.WaitForJobComplete) to check if job is Failed or Complete, and assert that it also has the corresponding FailureTarget, or SuccessCriteriaMet condition. WDYT?

mimowo replied:
Update: I have tentatively updated the helper functions for e2e and integration tests to assert that whenever we observe Failed or Complete conditions, then we also observe the corresponding interim conditions. PTAL.

mimowo commented:
/test pull-kubernetes-integration

mimowo commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -405,6 +405,12 @@ func getStatusValidationOptions(newJob, oldJob *batch.Job) batchvalidation.JobSt
+			AllowForSuccessCriteriaMetInExtendedScope:    true,
+		}
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
```
I see, so the case is `JobPodReplacementPolicy` (or `JobManagedBy`) is enabled and `SuccessCriteriaMet` is added, say for NonIndexedJob. Now, the features  `JobPodReplacementPolicy` (and `JobManagedBy`) get disabled, and with `JobSuccessPolicy` feature enabled all updates to the Job are rejected.
I fix this scenario and add the ""valid update of Job if SuccessCriteriaMet already present for NonIndexed Jobs; JobSuccessPolicy enabled, while JobManagedBy and JobPodReplacementPolicy disabled" test case

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -974,6 +974,27 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	return manageJobErr
 }
 
+func (jm *Controller) newFailureCondition(reason, message string) *batch.JobCondition {
```
This could work - it was my original idea, but then I thought that the functions `newFailureCondition` and `newSuccessCondition` have another benefit - they avoid duplication for setting the `ConditionTrue`, and `jm.clock.Now()`. So, maybe they have a right to exist even when the feature gates graduate. WDYT?

atiratree replied:
I would prefer less indirection after the graduation and duplicate the `ConditionTrue`, and `jm.clock.Now()`, but I suppose it is a matter of personal preference.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -1972,6 +1999,9 @@ func trackTerminatingPods(job *batch.Job) bool {
 	if feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
```
Done

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -1377,6 +1398,12 @@ func (jm *Controller) enactJobFinished(logger klog.Logger, jobCtx *syncJobCtx) b
+			logger.V(4).Info("Delaying marking the Job as finished, because there are still terminating pod(s)", "job", klog.KObj(job), "condition", jobCtx.finishedCondition.Type, "count", *jobCtx.terminating)
+			return false
+		}
+	}
```
I was rather thinking of ensuring this at the validation level, that when setting Terminal condition, `active=0, ready<=active, and terminating=0`, effectively re-adding the rules as in: https://github.com/kubernetes/kubernetes/pull/123792. when JobManagedBy is enabled. The advantage of validation is that it will ensure conformance for other Job controller implementations. WDYT?

mimowo replied:
Well, bugs can happen also when feature gate is GA, so we should think about it more.

I think these are main scenarios:
1. there is a bug in the Job controller which prevents adding terminal condition even though all pods are terminal, and don't have finalizers
2. there is a bug in Kubelet (see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase)
3. there is a bug in PodGC to transition it to Failed phase

For (1.) I believe we could just fix inside the controller, as with any other bug.
For (2.) or (3.) we may not have means to mitigate inside Job controller other than disabling the FG (but only until GA). 

I guess we need to decide if these (2.) and (3) cases are critical for the workflows. I would argue the consequence is rather limited, I think it only affects CronJobs which could prevent new job creation when concurrencyPolicy=Forbid. It can also affect external in-house controllers but they could adjust by using `failed || FailureTarget`.

alculquicondor replied:
I'm ok with what is implemented here and the API validation updates.

atiratree replied:
ack, api validation should be enough

smarterclayton replied:
It is not clear to me whether the resolution of this discussion was: "add additional validation that terminal conditions can't be set when `active > 0 || ready > active || terminating > 0`" or "no additional change is needed over what the PR currently implements"?

mimowo replied:
My reading of the resolution, which is currently implemented in was, add the following validation rules:
1. ready <= active (new rule in https://github.com/kubernetes/kubernetes/pull/125849 , see [here](https://github.com/kubernetes/kubernetes/blob/3c568d2d527f79b269f9433ebf9c6937a43d97cd/pkg/apis/batch/validation/validation.go#L586-L590))
2. terminal conditions can't be set when terminating > 0 (new rule in https://github.com/kubernetes/kubernetes/pull/125849 , see [here](https://github.com/kubernetes/kubernetes/blob/3c568d2d527f79b269f9433ebf9c6937a43d97cd/pkg/apis/batch/validation/validation.go#L581-L584))
3. terminal conditions can't be set when active > 0 (already on master, see [here](https://github.com/kubernetes/kubernetes/blob/27c6c30905107593845d4b31281f03fc8b39cc9f/pkg/apis/batch/validation/validation.go#L523-L525)

From 1 and 3 it follows also that terminal conditions can't be set when `ready>0`, so it does not need an explicit rule.

soltysh replied:
After recent changes this is now part of this PR, and that was my understanding as well.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -844,7 +844,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
+	if jobCtx.finishedCondition == nil && (feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) || delayTerminalCondition()) {
 		if failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget); failureTargetCondition != nil {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
 		} else if failJobMessage := getFailJobMessage(&job, pods); failJobMessage != nil {
```
The feature gate is also checked inside `getFailJobMessage` so `finishedCondition` would not be set in that case.

Still, to make it more readable we could refactor a bit: by analogy to `hasSuccessCriteriaMetCondition` we could have `hasFailureTargetCondition`. Then, we move `getFailJobMessage` down to the block [here](https://github.com/kubernetes/kubernetes/blob/ff8834cdd7abb9a2975f20dffb575d7f00a1d4d3/pkg/controller/job/job_controller.go#L855-L867). However, I would see this rather as a preparatory / follow up cleanup. WDYT?

mimowo replied:
+1 on removing the gated code, dropping the old branches would be a great simplification

alculquicondor replied:
In GA+2 is when we drop the feature gate, but the branching could be dropped earlier.

The reason we keep the feature gate is that an argument doing `--FooFeature=true` doesn't prevent the binary from starting.

dejanzele commented:
/retest

mimowo commented:
The `pull-kubernetes-conformance-kind-ga-only-parallel` failed because I tried to test that the interim failure conditions are set in the helper functions for e2e tests. However, the conformance tests also use these functions, and they would fail, because they only run on the GA features enabled (so this fix is disabled in GA).

mimowo commented:
@dejanzele @atiratree @alculquicondor I believe all the comments are addressed. PTAL. 

I also opened the follow up for validation: https://github.com/kubernetes/kubernetes/pull/125849. Let me know if you prefer these changes as one PR.

mimowo commented:
/retest

dejanzele commented on test/e2e/apps/job.go:
```diff
@@ -1186,22 +1238,26 @@ func waitForJobEvent(ctx context.Context, config watchEventConfig) {
-// waitForJobFailure uses c to wait for up to timeout for the Job named jobName in namespace ns to fail.
-func waitForJobFailure(ctx context.Context, c clientset.Interface, ns, jobName string, timeout time.Duration, reason string) error {
-	return wait.Poll(framework.Poll, timeout, func() (bool, error) {
+// waitForJobCondition uses c to wait for up to timeout for the Job named jobName in namespace ns to have the expected condition.
```
It doesn't relate to this PR, but the line `uses c to wait` doesn't  make much sense (even though `c` is the func arg)

mimowo replied:
Right, it seems like irrelevant technical detail. I dropped it and rephrased the comment a bit. PTAL.

dejanzele commented on test/integration/job/job_test.go:
```diff
@@ -1162,6 +1163,301 @@ func TestBackoffLimitPerIndex_JobPodsCreatedWithExponentialBackoff(t *testing.T)
 }
 
+// TestDelayTerminalPhaseCondition tests the fix for Job controller to delay
+// setting the terminal phase conditions (Failed and Complete) until all pods
```
nit: all other k8s resources are capitalised
```suggestion
// setting the terminal phase conditions (Failed and Complete) until all Pods
```

mimowo replied:
Done

dejanzele commented on test/integration/job/job_test.go:
```diff
@@ -1162,6 +1163,301 @@ func TestBackoffLimitPerIndex_JobPodsCreatedWithExponentialBackoff(t *testing.T)
+	succeedOnePodAndScaleDown := func(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1.Job) {
+		// mark one pod as succeeded
+		if err := setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodSucceeded, 0); err != nil {
+			t.Fatalf("Failed setting phase %q on Job Pod: %q", v1.PodSucceeded, err)
```
Somewhere %q formatter is used for the error and somewhere %v? Do we have any conventions around this?

mimowo replied:
I don't think we have any convention about that, thus the inconsistencies. We could have another PR to align it IMO

dejanzele replied:
Biggest issue which can happen here is that `%q` breaks error wrapping and `errors.Is` won't work, and devs will need to have deeper knowledge when checking for wrapped errors

mimowo replied:
Sure, I opened: https://github.com/kubernetes/kubernetes/pull/125914. PTAL

dejanzele commented:
/lgtm

alculquicondor commented:
/approve
/hold for last suggestions

mimowo committed:
Add more test cases for SuccessCriteriaMet

Cleanup error messages in the new code

Add validation for the Job controller fields

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -2046,6 +2076,110 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
+				},
+			},
+			wantFailedPodsMetric: 1,
+		},
```
Indeed, added the 3 new test cases, and also simplified the previously added 3 test cases for FailureTarget by removing the finalizer from the terminating pod already in setup.

dejanzele commented:
/lgtm

soltysh reviewed (approved):
/lgtm
/approve

/hold cancel

soltysh commented:
/triage accepted
/priority important-longterm

smarterclayton commented on test/e2e/apps/job.go:
```diff
@@ -83,6 +82,10 @@ var _ = SIGDescribe("Job", func() {
 		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
 		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
 
+		ginkgo.By("Awaiting for the job to have the interim success condition")
```
Is there an existing e2e test that validates that these conditions are not set until all pods are terminal?  Or was there a decision to test it only in unit and integration due to the lower likelihood that we could observe it?

While it tends to be harder to test, we have done it before for pods using a list/watch construct to accumulate the set of deltas and failed if the invariant is violated.  I suppose in this case we can't guarantee that pods and jobs have happens-before relationships between the two resource types across all possible cluster implementations.

mimowo replied:
> Is there an existing e2e test that validates that these conditions are not set until all pods are terminal? 

Not currently, but I'm open to discuss / add such a test.

> Or was there a decision to test it only in unit and integration due to the lower likelihood that we could observe it?

I think it wasn't discussed during the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#e2e-tests), so I'm not sure what others think, but I was a bit concerned with the gain / effort ratio. I think we don't really aim to test kubelet itself with Job test, so integrations tests are fairly good representative of what would happen. In particular, one needs to be careful about the ordering of events. I can imagine that when all pods transition to terminal it is technically possible, that the Job controller informer receives the event before test code receives it. As a result, there is a possibility that Job controller marks the Job failed, while the testing code still sees some of the pods not terminal. (didn't check this is a real possibility, sounds unlikely, but I don't see what can prevent it in principle other than timing).

What is relatively easy to do in e2e tests is that Job's `status.terminating=0` when the job is terminal, which is a Job-level proxy for all pods being terminal. I checked this manually by running the validation in https://github.com/kubernetes/kubernetes/pull/125849.

smarterclayton replied:
Agree the more complex test is necessary.

Verifying that terminating = 0 when the job final status is set would accomplish the objective of ensuring this behavior is both verified in conformant implementations and prevent regressions.  Is that captured in an existing e2e test or can we add it in this e2e?

mimowo replied:
Not captured currently, but easy to add. I don't think we need a dedicated new test for it. I think adding the check to one or a couple of existing e2e tests should be enough. WDYT?

mimowo replied:
Just added tentatively the asserts: https://github.com/kubernetes/kubernetes/pull/125510/commits/2da2618ca8d52b453f06415fec4774e7df924431. Two remarks:
1. to enable the asserts for successful Jobs (status.succeeded = spec.completions), I also now check the condition is Complete (otherwise there could be race conditions).
2. I don't add the asserts for conformance tests, because we only run them for GA feature gates, and the fix is behind beta. eventually, we will be able to verify all the checks in the helper function for Complete / Failed Jobs.

smarterclayton replied:
2 is fine.

mimowo replied:
Updated accordingly in the last commit

mimowo commented:
/retest

mimowo commented:
/test pull-kubernetes-integration
/test pull-kubernetes-verify 
/test pull-kubernetes-verify-lint
Failures due to overloaded infra

mimowo committed:
Verify the Job status active, ready and terminating are 0

Signed-off-by: Michal Wozniak <michalwozniak@google.com>

mimowo committed:
Delay setting terminal Job conditions until all pods are terminal

Fix the integration test typecheck

Fix after rebase

# Conflicts:
#	pkg/controller/job/job_controller_test.go

mimowo committed:
do not assert in e2e tests

alculquicondor commented:
/retest

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -823,9 +843,9 @@ done`}
-		ginkgo.By("Ensuring job exceed backofflimit")
 
-		err = waitForJobFailure(ctx, f.ClientSet, f.Namespace.Name, job.Name, e2ejob.JobTimeout, "BackoffLimitExceeded")
+		ginkgo.By("Ensuring job exceed backofflimit")
```
Can we check the other condition and other status fields here too?

mimowo replied:
Sure, I was thinking about this test too, added in the commit https://github.com/kubernetes/kubernetes/pull/125510/commits/0acffd6f2cfb668b031c0ca49c7ada892da17115

deads2k commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -568,7 +568,7 @@ func validateJobStatus(job *batch.Job, fldPath *field.Path, opts JobStatusValida
 		}
 	}
-	if ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) != batch.IndexedCompletion && isJobSuccessCriteriaMet(job) {
+	if !opts.AllowForSuccessCriteriaMetInExtendedScope && ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) != batch.IndexedCompletion && isJobSuccessCriteriaMet(job) {
```
I expected to see a validation change that corresponds to the verbiage here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated/README.md?plain=1#L419-L421

that indicates that jobs cannot be `Failed` or `Complete` unless `.status.terminating` is zero.

mimowo replied:
This is implemented under https://github.com/kubernetes/kubernetes/pull/125849 . I initially separated the PRs caused they touched different code. However, given how closely they are connected conceptually I think it makes sense to combine. 

I tentatively put them together by cherry-pick of the commit https://github.com/kubernetes/kubernetes/pull/125510/commits/70c49652704b3dba47b77888f53d34f20cbfcf0b.

PTAL @alculquicondor @soltysh @deads2k if you agree. Otherwise I will return to the idea of 2 PRs.

soltysh replied:
sgtm

mimowo committed:
Extend the e2e tests

mimowo commented:
/test pull-kubernetes-e2e-gce 
[Unrelated failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125510/pull-kubernetes-e2e-gce/1811692936509788160)

tenzen-y commented on pkg/controller/job/job_controller.go:
```diff
@@ -975,6 +975,27 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
+	if delayTerminalCondition() {
+		cType = batch.JobSuccessCriteriaMet
+	}
+	return newCondition(cType, v1.ConditionTrue, "", "", jm.clock.Now())
```
I'm sorry if I am missing any context and background.

I'm wondering if we should have a dedicated reason for these situations where the job doesn't have successPolicy, and the `deplayTerminalConfiton` is true. Because I guess that users can not understand the reason why job has `SuccessCriteriaMet` regardless of the job doesn't have any successPolicies.

What do you think?

mimowo replied:
I'm +1 on a dedicated reason for completing a job when `spec.completions == status.succeeded`, one natural name to consider would be `CompletionsReached`.

However. we didn't have the reason prior to this PR, so I don't consider it a regression (see [here[(https://github.com/kubernetes/kubernetes/blob/a39f42582f065de4b4708deef6709229f5c4d49b/pkg/controller/job/job_controller.go#L928)). I think we could add it as part of the SuccessPolicy KEP since it is still alpha / beta, but it would be a bit of an API-breaking change. WDYT @alculquicondor @soltysh @deads2k ?

soltysh replied:
I think it's safe and reasonable to add a reason explaining where the success criteria is coming from. I wouldn't consider that a breaking change, and can be easily addressed in followup or as part of the other work, as Michal mentioned.

tenzen-y replied:
That sounds reasonable to me. Thanks @mimowo @soltysh!
Actually, I'm working on the SuccessPolicy graduation PR. So, let me try it on the SuccessPolicy beta graduation PR. If we find any drawback to the dedicated reason in the graduation PR, we can give up to introduce the dedicated reason.

deads2k commented:
The validation and strategy changes lgtm.  I have not reviewed the controller changes.  Let me know on slack when the controller changes are lgtm/approved, I don't want to accidentally mark them.

mimowo commented:
> The validation and strategy changes lgtm. I have not reviewed the controller changes. Let me know on slack when the controller changes are lgtm/approved, I don't want to accidentally mark them.

@alculquicondor @soltysh could you renew your lgtm after the recent changes (including cherry-picking of the validation onto this branch)

soltysh reviewed (approved):
/lgtm
controller changes

deads2k commented:
The validation and strategy changes lgtm

/approve
